xbmc/games/dialogs/osd/DialogInGameSaves.cpp

```diff
@@ -0,0 +1,121 @@
+/*
+ *  Copyright (C) 2020 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DialogInGameSaves.h"
+#include "cores/RetroPlayer/playback/IPlayback.h"
+#include "cores/RetroPlayer/savestates/ISavestate.h"
+#include "cores/RetroPlayer/savestates/SavestateDatabase.h"
+#include "cores/RetroPlayer/guibridge/GUIGameRenderManager.h"
+#include "cores/RetroPlayer/guibridge/GUIGameSettingsHandle.h"
+#include "guilib/LocalizeStrings.h"
+#include "guilib/WindowIDs.h"
+#include "ServiceBroker.h"
+#include "settings/GameSettings.h"
+#include "settings/MediaSettings.h"
+#include "utils/log.h"
+//#include "utils/Variant.h"
+
+using namespace KODI;
+using namespace GAME;
+using namespace RETRO;
+
+CDialogInGameSaves::CDialogInGameSaves()
+    : CDialogGameVideoSelect(WINDOW_DIALOG_IN_GAME_SAVES)
+{
+}
+
+std::string CDialogInGameSaves::GetHeading()
+{
+  return g_localizeStrings.Get(35249); // "Saved games"
+}
+
+void CDialogInGameSaves::PreInit()
+{
+  m_items.Clear();
+
+  InitSavedGames();
+
+  CFileItemPtr item = std::make_shared<CFileItem>(g_localizeStrings.Get(223)); // "New"
+  item->SetIconImage("DefaultAddSource.png");
+  item->SetPath("");
+
+  m_items.AddFront(std::move(item), 0);
+}
+
+void CDialogInGameSaves::InitSavedGames()
+{
+  auto gameSettings = CServiceBroker::GetGameRenderManager().RegisterGameSettingsDialog();
+
+  // save current game
+  IPlayback* playback = gameSettings->GetPlayback();
+  playback->CreateSavestate();
+
+  CSavestateDatabase db;
+  db.GetSavestatesNav(m_items, gameSettings->GetPlayingGame(), gameSettings->GameClientID());
```

Comment:
`CSavestateDatabase::GetSavestatesNav()` is not idempotent; if called multiple times, `m_items` will change each time. On initialization, we want `m_items` to equal the savestates only, not include the savestates from the last call. We have two options:

1. Clear `m_items` directly before this
2. Clear `m_items` in a deinitialization function

If we take step 2, then CDialogInGameSaves won't have an invalid state when not initialized. Therefore, immediately after this function, I expect to see a deinitialization function with the line `m_items.Clear()`.

Reply:
I clear m_items on exit here https://github.com/garbear/xbmc/pull/120/files/0989016af62b738e594788557af2cf76cc5838fe#diff-ee3b0edb525fed145407f14fd64ec6edR98-R101